### ISSUE TITLE
More advanced architectures

### ISSUE THREAD

**Issue by Zahra-Alavi** [Zahra-Alavi/ESMfluc#27]
- Implement bilstm + attn + residual 
- implement multi-task learning: classification + regression

**Comment by ntran18** [comment]
I implemented the regression model in `mdcath_pipeline` folder in `mdcath-data-generation` branch. The pipeline works both for atlas and mdcath. 

After trying and trying, I got a model that trained on 450 K mdcath.

Model specs:
- ESM model as a backbone + 2 linear layers (first linear layer compressed 320-dimensional to 180-dimensional hidden state, the last linear layer compressed 180-dimensional hidden state into 1 single value) + Relu and Dropout (rate=0.1) in between two linear layers.

- Loss function: Weight MSE. If the expected neq value is >10, then we add weight_factor to the loss to prioritize accuracy on highly flexible residues.

- ESM model used: `esm2_t6_8M_UR50D`
- Command line I used to train:
```
CUDA_VISIBLE_DEVICES=0 python train.py --batch_size 32 --weight_factor 5 --lr 2e-5 --epochs 20 --batch_size 32 --num_unfreeze_layers 6 --temperatures 450 --weight_threshold 10
```

- The current model is saved in the checkpoint folder name: `checkpoints/esm-fluc-epoch=11-val_loss=6.97.ckpt`

Here is the result:

Evaluation Temp | Spearman | Pearson | MAE | Status
-- | -- | -- | -- | --
320K | 0.3417 | 0.3623 | 4.7146 | Unseen (Generalization)
348K | 0.3894 | 0.4189 | 4.3884 | Unseen (Generalization)
379K | 0.4617 | 0.4985 | 3.8164 | Unseen (Generalization)
413K | 0.5959 | 0.6077 | 2.8390 | Unseen (Generalization)
450K | 0.7132 | 0.7134 | 1.8034 | Training Set
atlas | 0.2132 | 0.2311 | 4.2590 | Unseen (Generalization)

Raw pictures:
For 320K
<img width="355" height="130" alt="Image" src="https://github.com/user-attachments/assets/dfa47e50-2222-4fdd-8dda-b8ad535db295" />

For 348 K:

<img width="353" height="130" alt="Image" src="https://github.com/user-attachments/assets/ccfc853b-629c-487f-8f99-ea2bd97e610d" />

For 379K:
<img width="355" height="131" alt="Image" src="https://github.com/user-attachments/assets/ff77c040-312a-4c17-b19b-0545aaefe3da" />



For 413K:

<img width="352" height="132" alt="Image" src="https://github.com/user-attachments/assets/58dfbf44-2321-4002-bed5-218b47b5f171" />

For 450K:

<img width="353" height="130" alt="Image" src="https://github.com/user-attachments/assets/0bdf42b0-c403-495a-a771-16e4a9f79503" />

For atlas:

<img width="352" height="111" alt="Image" src="https://github.com/user-attachments/assets/08cd78cb-4a06-442b-a746-a2d0914c3848" />

**Comment by ntran18** [comment]
The pipeline is also designed to train multiple temperatures at once. If there are multiple temperatures, after getting the embeddings from ESM, the model will add one extra column for temperature. The temperature is normalized using the min-max formula.

For example, `esm2_t6_8M_UR50D` have shape [batch_size, sequence_len, 320]. The embedding dimension is 320 + 1 (for temperature) then it will feed into the regression head.
